### ISSUE TITLE
change Atomic App Base Image to use ONBUILD

### ISSUE THREAD

**Issue by goern** [projectatomic/atomicapp#195]
transforming https://github.com/projectatomic/atomicapp/issues/142#issuecomment-113213570 into an issue

# Change the atomicapp container to use ONBUILD and standard locations

The atomicapp container currently uses /application-entity as the runtime location. Downstream containers add application information and artifacts in their Dockerfiles. That location is modifiable by changing where application data gets stored and overriding the LABELs for run and install. I propose we mandate that location and not allow overrides by containers built from the atomicapp image.

The ONBUILD Dockerfile directive allows for triggered actions to be taken on the creation of a child image from a parent, e.g. a Nulecule app from the atomicapp container. If we standardize the locations of application added information, we can moved the ADD directives that currently exist in the child Nulecule app Dockerfiles to the parent atomicapp Dockerfile.

`ONBUILD ADD . /application-entity/`

When a child app is built, docker will then run the ADD directives and deliver the local application content to the correct container locations.

```
Sending build context to Docker daemon 
Step 0 : FROM nzwulfin/atomicapp
# Executing 2 build triggers
Trigger 0, ADD . /application-entity/
Step 0 : ADD . /application-entity/
Trigger 1, ADD /artifacts /application-entity/artifacts
Step 0 : ADD /artifacts /application-entity/artifacts
```

**Comment by aweiteka** [comment]
Reference work from @nzwulfin:
 https://github.com/nzwulfin/atomicapp/tree/onbuild
https://github.com/nzwulfin/nulecule/tree/onbuild

**Comment by aweiteka** [comment]
Thanks for writing this up. One note: this may break all images build from atomicapp image. We have no way to contact people who are downstream. Probably best to just do it since it's early in the project.

**Comment by aweiteka** [comment]
I'm confused about `/atomicapp` vs `/application-entity`. It appears the *working* directory is /atomicapp. But then the actual working directory is /atomicapp/.workdir/...

/application-entity is just where downstream projects put their stuff but it's installed to /atomicapp. Can't the working directory just be a tmpdir?

We need to get this sorted out in relation to what it means to install something. @vpavlin any comments on the current design or how we can simplify this?

**Comment by dustymabe** [comment]
@vpavlin ^^

**Comment by vpavlin** [comment]
Problems with ONBUILD approach, which I already stated earlier, are that

1. You will get random stuff from current directory (or any directory you point docker build at) into the image.
2. We would need to standardise `./artifacts` directory name (which we wanted to avoid, IIRC)

@aweiteka Working directory for `install` is `/atomicapp` - i.e. to unpack files from container. Working directory for `run` is `/atomicapp/.workdir` - i.e. to store substituted/templated files.

**Comment by nzwulfin** [comment]
@vpavlin  Agree,  the original ONBUILD approach was looking at publishing an atomicapp based container (ISV style) rather than continued and further use by an end-user.  The publisher would need to recognize that everything would get pulled in and have a clean final working directory to build a final container.  

As far as the directory structure, I was working off the existing Dockerfiles,  so those could be standardized to whatever. I didn't (and really still don't) know enough about the directory structures  that atomicapp uses, I was just using what I knew worked for the helloapache-app container.

**Comment by vpavlin** [comment]
@nzwulfin There is no directory structure defined. I wanted to have a standard directory for artifacts originaly, but @goern and @aweiteka explained it's bad and we should enforce any names apart from `Nulecule`. The spec was change to accommodate this and you now need to specify a full path to artifact in Nulecule file. That's why I am not a big fan of this proposal for using `ONBUILD` for `ADD`ing content into the image. But if others are fine with that, I won't block it:)
